File: bun-typescript/src/diff.ts
import fs from 'node:fs';
import { DiffResult, ParsedTask, TASK_FILE } from './schema';

// Function to read and parse a JSON file
function read(path: string) {
    const data = fs.readFileSync(path, 'utf8');
    const result = TASK_FILE.safeParse(JSON.parse(data));
    if (!result.success) throw new Error('Invalid JSON');
    return result.data;
}



// Read the base and new task data
const base_tasks = read('./output-base.json');
const new_tasks = read('./output-new.json');

const diffs: DiffResult[] = [];

const extract_task = (task: ParsedTask) => {
    const { text, line, file } = task;
    return { text, line, file };
}

const same_text = (task1: ParsedTask, task2: ParsedTask) => {
    // return true if the tasks have the same text
    return task1.text.trim() === task2.text.trim();
}

// process the tasks
new_tasks.forEach((new_task: ParsedTask) => {
    // first we want to check if we can find a task in the base tasks that has the same text but different line number or file
    const same_text_result = base_tasks.find(bt => same_text(bt, new_task));
    if (same_text_result) {
        if (same_text_result.line !== new_task.line || same_text_result.file !== new_task.file) {
            const data = { old: extract_task(same_text_result), new: extract_task(new_task) };
            diffs.push({ id: same_text_result.id, tag: new_task.tag, type: 'MOVE', data });
        } else {
            const data = { old: extract_task(same_text_result), new: extract_task(new_task) };
            diffs.push({ id: same_text_result.id, tag: new_task.tag, type: 'SAME', data });
        }
        return;
    }
    // then if the line number and tag are the same but the text is different
    const same_line_result = base_tasks.find(bt => bt.line === new_task.line && bt.tag === new_task.tag);
    if (same_line_result && !same_text(same_line_result, new_task)) {
        const data = { old: extract_task(same_line_result), new: extract_task(new_task) };
        diffs.push({ id: same_line_result.id, tag: new_task.tag, type: 'UPDATE', data });
        return;
    }

    // if we get here then we have a new task
    const data = { old: null, new: extract_task(new_task) };
    diffs.push({ id: new_task.id, tag: new_task.tag, type: 'NEW', data });    
});

base_tasks.forEach(base_task => {
    // check if the task in the new tasks, and if not add a 'DELETE' diff
    const in_diffs = diffs.find(d => d.id === base_task.id);
    if (!in_diffs) {
        const data = { old: extract_task(base_task), new: null };
        diffs.push({ id: base_task.id, tag: base_task.tag, type: 'DELETE', data });
    }
});

console.log(JSON.stringify(diffs, null, 2));

File: go/diff.go
package main

import (
	"encoding/json"
	"os"
	"strings"
)

func read(path string) ([]ParsedTask, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var tasks []ParsedTask
	err = json.Unmarshal(data, &tasks)
	if err != nil {
		return nil, err
	}

	return tasks, nil
}

func extractTask(task ParsedTask) *DiffInfo {
	return &DiffInfo{
		Text: task.Text,
		Line: task.Line,
		File: task.File,
	}
}

func sameText(task1, task2 ParsedTask) bool {
	return strings.TrimSpace(task1.Text) == strings.TrimSpace(task2.Text)
}

func generate_diff(dir1 string, dir2 string) ([]DiffResult, error) {
	base_tasks, err := read(dir1)
	if err != nil {
		return nil, err
	}

	new_tasks, err := read(dir2)
	if err != nil {
		return nil, err
	}

	var diffs []DiffResult

	for _, new_task := range new_tasks {
		found := false

		// search for the same text but different line or file
		for _, base_task := range base_tasks {
			if sameText(base_task, new_task) {
				diffType := "SAME"
				if base_task.Line != new_task.Line || base_task.File != new_task.File {
					diffType = "MOVE"
				}

				diffs = append(diffs, DiffResult{
					ID:   base_task.ID,
					Tag:  new_task.Tag,
					Type: diffType,
					Data: DiffData{
						Old: extractTask(base_task),
						New: extractTask(new_task),
					},
				})
				found = true
				break
			}
		}
		if found {
			continue
		}

		// then we need to see if line number and tag are same but text is different
		for _, base_task := range base_tasks {
			if base_task.Line == new_task.Line && base_task.Tag == new_task.Tag && !sameText(base_task, new_task) {
				diffs = append(diffs, DiffResult{
					ID:   base_task.ID,
					Tag:  new_task.Tag,
					Type: "UPDATE",
					Data: DiffData{
						Old: extractTask(base_task),
						New: extractTask(new_task),
					},
				})
				found = true
				break
			}
		}
		if found {
			continue
		}

		diffs = append(diffs, DiffResult{
			ID:   new_task.ID,
			Tag:  new_task.Tag,
			Type: "NEW",
			Data: DiffData{
				Old: nil,
				New: extractTask(new_task),
			},
		})
	}

	for _, base_task := range base_tasks {
		found := false
		for _, diff := range diffs {
			if diff.ID == base_task.ID {
				found = true
				break
			}
		}

		if !found {
			diffs = append(diffs, DiffResult{
				ID:   base_task.ID,
				Tag:  base_task.Tag,
				Type: "DELETE",
				Data: DiffData{
					Old: extractTask(base_task),
					New: nil,
				},
			})
		}
	}

	return diffs, nil
}
